File: Marriage/MarriageApi/Marriage.Application/Impl/MarriageStatus.cs
using Marriage.Entity.Application;
using Marriage.Entity.Application.MarriageStatus;
using System;
using System.Collections.Generic;
using System.Text;

namespace Marriage.Application.Impl
{
    /// <summary>
    /// 相亲状态
    /// </summary>
    public class MarriageStatus : BaseAction, IMarriageStatus
    {
        public Domain.IMatchmaker _Matchmaker { get; set; }
        public Domain.IMarriageArrange _MarriageArrange { get; set; }

        /// <summary>
        /// 获取相亲状态
        /// </summary>
        public GetMarriageStatusResponse GetMarriageStatus(GetMarriageStatusRequest request)
        {
            string title = "获取相亲状态";
            string requestContent = Utility.Common.ToJson(request);
            GetMarriageStatusResponse response = new GetMarriageStatusResponse();

            this.InitMessage();

            this.IsNullRequest(request, response);

            //1、获取用户信息
            int stepNo = 1;
            this.GetMatchmakerById(stepNo, request.LoginUserId, response);

            //2、获取相亲安排
            stepNo += 1;
            var marriageArrange = GetViewMarriageArrange(stepNo, request.MarriageArrangeId, response);

            //2、获取相亲状态
            stepNo += 1;
            GetMarriageStatus(stepNo, marriageArrange, request, response);

            //3、执行结束
            this.ExecEnd(response);

            //日志记录
            return this.SetReturnResponse<GetMarriageStatusResponse>(title, "GetMarriageStatus", requestContent, response);
        }

        /// <summary>
        /// 保存相亲状态
        /// </summary>
        public SaveMarriageStatusResponse SaveMarriageStatus(SaveMarriageStatusRequest request)
        {
            string title = "保存相亲状态";
            string requestContent = Utility.Common.ToJson(request);
            SaveMarriageStatusResponse response = new SaveMarriageStatusResponse();

            this.InitMessage();

            this.IsNullRequest(request, response);

            //1、获取用户信息
            int stepNo = 1;
            this.GetMatchmakerById(stepNo, request.LoginUserId, response);

            //2、获取相亲安排
            stepNo += 1;
            var marriageArrange = GetViewMarriageArrange(stepNo, request.MarriageArrangeId, response);

            //2、保存相亲状态
            stepNo += 1;
            SaveMarriageStatus(stepNo, marriageArrange, request, response);

            //3、执行结束
            this.ExecEnd(response);

            //日志记录
            return this.SetReturnResponse<SaveMarriageStatusResponse>(title, "SaveMarriageStatus", requestContent, response);
        }

        /// <summary>
        /// 保存相亲状态
        /// </summary>
        /// <param name="stepNo"></param>
        /// <param name="marriageArrange"></param>
        /// <param name="request"></param>
        /// <param name="response"></param>
        /// <returns></returns>
        private bool SaveMarriageStatus(int stepNo, Entity.Domain.MarriageArrange marriageArrange, SaveMarriageStatusRequest request, IResponse response)
        {
            Func<bool> execStep = () =>
            {
                Guid loginUserId = Guid.Parse(request.LoginUserId);

                if (marriageArrange.AppMatchmakerId != loginUserId) return false;

                Entity.Domain.MarriageArrange entity = new Entity.Domain.MarriageArrange();

                entity.BookMarryDate = GetStringDate(request.BookMarryDate);
                entity.BreakUpDate = GetStringDate(request.BreakUpDate);
                entity.BreakUpReason = request.BreakUpReason;
                entity.CancelReason = request.CancelReason;
                entity.IsManAgree = request.IsManAgree;
                entity.IsWomanAgree = request.IsWomanAgree;
                entity.MarriageArrangeId = request.MarriageArrangeId;
                entity.MarryDate = GetStringDate(request.MarryDate);
                entity.NoManAgreeRemark = request.NoManAgreeRemark;
                entity.UpdateUser = loginUserId;
                entity.NoWomanAgreeRemark = request.NoWomanAgreeRemark;
                entity.Status = request.Status;

                return _MarriageArrange.SaveMarriageStatus(entity);
            };

            return this.UpdateEntityData(stepNo, "保存相亲状态", "GetMarriageArrange", response, execStep);
        }

        DateTime GetStringDate(string date)
        {
            if (string.IsNullOrEmpty(date)) return DateTime.MinValue;

            return DateTime.Parse(date);
        }

        /// <summary>
        /// 以主键获取相亲安排
        /// </summary>
        /// <param name="stepNo"></param>
        /// <param name="marriageArrangeId"></param>
        /// <param name="response"></param>
        /// <returns></returns>
        private Entity.Domain.MarriageArrange GetViewMarriageArrange(int stepNo, Guid marriageArrangeId, IResponse response)
        {
            Func<Entity.Domain.MarriageArrange> execStep = () =>
            {
                return _MarriageArrange.GetViewMarriageArrange(marriageArrangeId);
            };

            return this.GetEntityData<Entity.Domain.MarriageArrange>(stepNo, "以主键获取相亲安排", "GetMarriageArrange", response, execStep);
        }

        /// <summary>
        /// GetMarriageStatus
        /// </summary>
        /// <param name="stepNo"></param>
        /// <param name="marriageArrange"></param>
        /// <param name="request"></param>
        /// <param name="response"></param>
        /// <returns></returns>
        private Entity.Domain.MarriageArrange GetMarriageStatus(int stepNo, Entity.Domain.MarriageArrange marriageArrange, GetMarriageStatusRequest request, GetMarriageStatusResponse response)
        {
            Func<Entity.Domain.MarriageArrange> execStep = () =>
            {
                Guid loginUserId = Guid.Parse(request.LoginUserId);

                if (marriageArrange.AppMatchmakerId != loginUserId) return null;

                response.BookMarryDate = GetDateString(marriageArrange.BookMarryDate);
                response.BreakUpDate = GetDateString(marriageArrange.BreakUpDate);
                response.MarryDate = GetDateString(marriageArrange.MarryDate);
                response.BreakUpReason = marriageArrange.BreakUpReason;
                response.CancelReason = marriageArrange.CancelReason;
                response.IsManAgree = marriageArrange.IsManAgree;
                response.IsWomanAgree = marriageArrange.IsWomanAgree;
                response.MarriageArrangeId = marriageArrange.MarriageArrangeId;
                response.NoManAgreeRemark = marriageArrange.NoManAgreeRemark;
                response.NoWomanAgreeRemark = marriageArrange.NoWomanAgreeRemark;
                response.Status = marriageArrange.Status;

                return marriageArrange;
            };

            return this.GetEntityData<Entity.Domain.MarriageArrange>(stepNo, "以用户获取相亲安排", "GetMarriageArrangeByUser", response, execStep);
        }

        string GetDateString(DateTime date)
        {
            if (date == DateTime.MinValue) return null;
            return date.ToString("yyyy-MM-dd");
        }

        /// <summary>
        /// 以主键获取红娘信息
        /// </summary>
        /// <param name="stepNo"></param>
        /// <param name="userId"></param>
        /// <param name="response"></param>
        /// <returns></returns>
        private Entity.Domain.Matchmaker GetMatchmakerById(int stepNo, string userId, IResponse response)
        {
            Func<Entity.Domain.Matchmaker> execStep = () =>
            {
                var entity = _Matchmaker.GetMatchmakerById(Guid.Parse(userId));
                if (entity == null || entity.Status != 1) return null;
                return entity;
            };

            return this.GetEntityData<Entity.Domain.Matchmaker>(stepNo, "以主键获取红娘信息", "GetMatchmakerInfoById", response, execStep);
        }
    }
}

File: Marriage/MarriageMatchmakerH5/src/routes/marriage/marriageStatus.js
import { EntityPageEdit } from "PageTemplates";
import { Common } from 'UtilsCommon';

export default EntityPageEdit("marriage_marriageStatus", "MarriageStatus", document.title + '-相亲状态', 1800, {
    expandSetEntityData({ entityData, props, view }) {
        let message = '';
        const { Status } = entityData;
        if (Status === 0 || Status === 1 || Status === 3) {
            entityData.IsManAgree = 0;
            entityData.NoManAgreeRemark = '';
            entityData.IsWomanAgree = 0;
            entityData.NoWomanAgreeRemark = '';
            entityData.CancelReason = '';
            entityData.BookMarryDate = null;
            entityData.MarryDate = null;
            entityData.BreakUpDate = null;
            entityData.BreakUpReason = '';
        }
        else if (Status === 2) {
            entityData.CancelReason = '';
            entityData.BookMarryDate = null;
            entityData.MarryDate = null;
            entityData.BreakUpDate = null;
            entityData.BreakUpReason = '';

            if (entityData.IsManAgree === 0 && Common.isNullOrEmpty(entityData.NoManAgreeRemark)) {
                message = "当男方不同意时，需输入不同意原因！";
            }
            else if (entityData.IsWomanAgree === 0 && Common.isNullOrEmpty(entityData.NoWomanAgreeRemark)) {
                message = "当女主不同意时，需输入不同意原因！";
            }
            else if (entityData.IsManAgree === 1 && entityData.IsWomanAgree === 1) {
                message = "当状态为无意向时，男女双方不能选择都同意！";
            }
            else {
                if (entityData.IsManAgree) entityData.NoManAgreeRemark = '';
                if (entityData.IsWomanAgree) entityData.NoWomanAgreeRemark = '';
            }
        }
        else if (Status === 4) {
            entityData.IsManAgree = 0;
            entityData.NoManAgreeRemark = '';
            entityData.IsWomanAgree = 0;
            entityData.NoWomanAgreeRemark = '';
            entityData.CancelReason = '';
            entityData.BreakUpDate = null;
            entityData.BreakUpReason = '';

            if (Common.isNullOrEmpty(entityData.BookMarryDate)) {
                message = "请选择订婚日期！";
            }
        }
        else if (Status === 5) {
            entityData.IsManAgree = 0;
            entityData.NoManAgreeRemark = '';
            entityData.IsWomanAgree = 0;
            entityData.NoWomanAgreeRemark = '';
            entityData.CancelReason = '';
            entityData.BreakUpDate = null;
            entityData.BreakUpReason = '';

            if (Common.isNullOrEmpty(entityData.MarryDate)) {
                message = "请选择结婚日期！";
            }
        }
        else if (Status === 6) {
            entityData.IsManAgree = 0;
            entityData.NoManAgreeRemark = '';
            entityData.IsWomanAgree = 0;
            entityData.NoWomanAgreeRemark = '';
            entityData.CancelReason = '';

            if (Common.isNullOrEmpty(entityData.BreakUpDate)) {
                message = "请选择分手日期！";
            }
            else if (Common.isNullOrEmpty(entityData.BreakUpReason)) {
                message = '请输入分手原因！'
            }
        }
        else if (Status === 7) {
            if (Common.isNullOrEmpty(entityData.CancelReason)) {
                message = '请输入取消原因！'
            }
        }

        if (message) {
            this.alert(message);
            return false;
        }
        return entityData;
    },
});